packages/react/src/hooks/useVisibilityLoader.js

```diff
@@ -0,0 +1,88 @@
+import { useEffect, useRef, useState } from 'react';
+import warning from 'warning';
+import { merge } from 'lodash-es';
+
+import { browserSupports } from '../utils/componentUtilityFunctions';
+
+const defaultOptions = {
+  /** are there more items available to load */
+  hasMoreToLoad: true,
+  /** is the application already loading new data */
+  isLoading: false,
+  /** a callback that will be triggered when the ref element becomes visible */
+  onVisible: () => {},
+  /* options passed to IntersectionObserver, see: https://developer.mozilla.org/en-US/docs/Web/API/Intersection_Observer_API */
+  intersectionObserverOptions: {
```

Comment:
Not a fan of exposing the intersection options.  The fact that we are using intersection observer should be private, and we should craft a set of parameters that we end up rewriting for the intersection observer vs exposing the entire "options" to the developer and forcing them to understand what an intersection object is and how it works, etc.  Sadly even after I read the document for intersection observer it's not clear to me what rootMargin is doing :(

Reply:
That's a good point. I can simply that so we're not using all IntersectionObserver options. It may not be necessary to expose any of them, but I included them to give devs more flexibility. By exposing the rootMargin for example, it gives them control over how large the "visible" area is and how many items are being "pre-loaded" off-screen as the user is scrolling.

Yes, the docs are extremely unclear. A positive margin expands the intersection area outside the bounds of the root object, so the callback is trigger sooner where a negative margin shrinks the area inside the bounds of the root element. So, `100px 0px 100px 0px' is expanding the area 100px above and 100px below the root area, so that the element is marked as intersecting 100px sooner than it would if the margin were "0px" (the default).